### PR TITLE
Flow Layout Container implementation

### PR DESCRIPTION
**Pull request by singhmanp0707**
# Support for Flow Layout Container

- This PR adds support for the Flow Layout container for elements of type AdaptiveCard, Container, Column and Table Cell.
- This PR introduces a feature flag resolver interface that allows the host to implement and register it, enabling control of features through ECS settings.

A new **_layouts_** property is introduced on container elements. The value is an array of Layout objects. The layout objects look like this:

Screenshot 2024-08-23 at 5 03 23 PM

The layout to apply is determined the following way: 

- Look for the first layout in the list whose targetWidth property matches the current host width. If one is found, apply it 

- Otherwise, look for a layout that doesn’t have a targetWidth defined. If one is found, apply it 

- Otherwise, apply a default Layout.Stack layout and ignore other layouts in the list 

# Spec for Layout.Flow

Screenshot 2024-08-23 at 4 57 38 PM

Sample Card:

Media (1)

Media (2)

### CODE REVIEW

**Review by ardlank — APPROVED** [review]

**Review by wyAbhishek — APPROVED** [review]

**Review by prpatwa — CHANGES_REQUESTED** [review]

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/CompoundButtonRenderer.java [hunk 1]
```diff
@@ -213,7 +224,7 @@ public Drawable createCustomDrawable(Context context, String backgroundColor, in
     }
 
     // Helper method to convert dp to pixels
-    private int dpToPx(Context context, int dp) {
```
[the remark this answers is not in the document]
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/CompoundButtonRenderer.java]: done

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/Util.java [hunk 2]
```diff
@@ -655,9 +796,37 @@ static long getSizeClosestToGivenSize(List<Long> availableSizes, Long targetIcon
         return closestSize;
     }
 
+    public static boolean isFlowLayoutEnabled() {
```
[the remark this answers is not in the document]
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/Util.java]: moved to a new file.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/AdaptiveCardRenderer.java [hunk 3]
```diff
@@ -220,6 +246,15 @@ public View internalRender(RenderedAdaptiveCard renderedCard,
         return rootLayout;
     }
 
+    private static FlexboxLayout getFlexboxContainerForLayout(Context context) {
+        FlexboxLayout flexboxLayout = new FlexboxLayout(context);
+        flexboxLayout.setFlexDirection(FlexDirection.ROW);
+        flexboxLayout.setFlexWrap(FlexWrap.WRAP);
+        flexboxLayout.setAlignItems(AlignItems.FLEX_START);
```
[the remark this answers is not in the document]
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/AdaptiveCardRenderer.java]: AlignItems is the alignment of the items on the flex line across cross axis. The default value is FLEX_START. Yes, it needs to be FLEX_START always irrespective of the horizontal alignment. Horizontal alignment is done using justifyContent property.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ColumnRenderer.java [hunk 4]
```diff
@@ -179,6 +194,27 @@ public View render(
         return columnLayout;
     }
 
+    private static ViewGroup getAppropriateContainerForLayout(Context context, Layout layoutToApply, Column column) {
+        ViewGroup layoutContainer;
+        if (layoutToApply.GetLayoutContainerType() == LayoutContainerType.Flow) {
+            FlexboxLayout flexboxLayout = new FlexboxLayout(context);
+            flexboxLayout.setFlexDirection(FlexDirection.ROW);
+            flexboxLayout.setFlexWrap(FlexWrap.WRAP);
+            flexboxLayout.setAlignItems(AlignItems.FLEX_START);
```
[the remark this answers is not in the document]
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ColumnRenderer.java]: same as per earlier comment.
Removed the `flexboxLayout.setAlignItems(AlignItems.FLEX_START);` as the default value is this only.

**Review by prpatwa — APPROVED** [review]

**Comment by GabrielMedAlv** [comment]
Is there some unit test could be added for this feature? If a card with nested layout could be added to ensure is working would be great.

**Review by GabrielMedAlv — APPROVED** [review]

**Comment by singhmanp0707** [comment]
> Is there some unit test could be added for this feature? If a card with nested layout could be added to ensure is working would be great.

The NestedFlowLayout example is already added in the samples.

**Pull request merged**